### PR TITLE
ci: add support for `/retest` command

### PR DESCRIPTION
**Pull request by jpsim**
Commenting `/retest` on a PR will re-run failed jobs in its GitHub Actions workflows.

image

Copied from https://github.com/containernetworking/plugins.

### CODE REVIEW

**Comment by jpsim** [comment]
The `/retest` command will only work once this PR is merged, so I can't test it here yet.

**Comment by jpsim** [comment]
Right now, this will re-run all CI jobs, including ones that are already succeeding. I might look into making it just re-run failing jobs at a later time.

**Comment by jpsim** [comment]
In my testing, it takes 20 to 30 seconds for the command to take effect: https://github.com/jpsim/ZenTuner/actions/workflows/commands.yml

**Comment by keith** [comment]
there is some custom stuff on envoy that we could reuse here probably

**Comment by jpsim** [comment]
> there is some custom stuff on envoy that we could reuse here probably

i looked at that but Envoy’s CI is all Azure Pipelines so it doesn’t really carry over.

**Comment by mattklein123** [comment]
cc @itayd who helps us with RepoKitteh in the main repo and is working on "something better" which we can hopefully use here also when it is ready.

**Comment by itayd** [comment]
don't we already have retest working here for azp?

**Comment by jpsim** [comment]
Envoy Mobile uses GitHub Actions for all of its CI, there are no azp jobs. There's also [one CircleCI job](https://github.com/envoyproxy/envoy-mobile/blob/main/.circleci/config.yml) that deploys docs when new versions are pushed.

**Comment by Augustyniak** [comment]
Speaking from the perspective of a person who did a few contributions in EM repo without being a maintainer I think that having a `/retest` command is a quality of life improvement - especially if we find a way to eventually make it smarter so that it does *not* rerun green jobs (that would've been pretty awesome). 

Pretty often I found myself pushing multiple empty commits to a branch just to get rid jobs that decided to go red randomly. Having `/retest` would've allowed me to have to do less of a git work.

**Comment by goaway** [comment]
Now that GitHub Actions directly supports rerunning failed jobs only, I think that's almost always preferable to rerunning all of CI. We have finite resources, and CI jobs do get backed up. If this could be improved to rerun failed jobs only, I feel it would be a lot more compelling. That said, for the occasions where rerunning everything really is what you want (post-CI outage?), I don't see the harm in having this.

**Comment by jpsim** [comment]
Thanks for the feedback everyone. After spending some time looking at the API responses, and looking at GitHub's REST API documentation, I found an endpoint to rerun failed jobs from a workflow run.

Unlike `rerun_url`, this endpoint isn't in the run API response, but we can construct it manually easily.

Documentation for it is here: https://docs.github.com/en/rest/actions/workflow-runs#re-run-failed-jobs-from-a-workflow-run

I've tested it on a separate repo and it worked as expected: https://github.com/jpsim/ZenTuner/pull/47

**Comment by jpsim** [comment]
@goaway @Augustyniak did my latest updates address your concerns about only re-running failing jobs?

File: .github/actions/retest-action/action.yml
Comment (goaway) [inline .github/actions/retest-action/action.yml:2]: Let's update this comment to reflect that the new behavior will specifically rerun failing jobs.

**Comment by goaway** [comment]
Thanks @jpsim! This is fantastic, and should be a huge improvement over needing to navigate the UI to restart each failing individually - in addition to solving the permissions issue for external contributors.

**Review by Augustyniak — APPROVED** [review]

**Pull request merged**
